### PR TITLE
Suggest using block for `extern "abi" fn` with no body

### PR DESCRIPTION
**Pull request by aDotInTheVoid**
@rustbot modify labels: +A-diagnostics

### CODE REVIEW

**Review by nagisa — COMMENTED** [review]
Lets run a perfbot and see if spans in extern are reasonably cheap addition.

**Comment by aDotInTheVoid** [comment]
Updated the diagnostic messages. Perf result is unfortunate, not sure if it's Ok to land like this, if it's possible to get the better errors without the perf overhead, or how to decide if it's worth it.

**Comment by nagisa** [comment]
@nnethercote I wonder what your take is here. Do you think there might be a better place to keep the `Span` for `extern` here?

File: compiler/rustc_ast_passes/src/ast_validation.rs [hunk 1]
```diff
@@ -1190,8 +1204,38 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
...
+                            } else {
+                                "extern {".to_owned()
+                            };
+
+                            let end_suggestion = " }".to_owned();
+                            let end_span = item.span.shrink_to_hi();
+
+                            e
+                            .multipart_suggestion(
+                                "if you meant to declare an externaly defined function, use an `extern` block",
```
Comment (nnethercote) [inline compiler/rustc_ast_passes/src/ast_validation.rs:1223]:
```suggestion
                                "if you meant to declare an externally defined function, use an `extern` block",
```

**Comment by nnethercote** [comment]
The performance regressions are very small and on only a couple of benchmarks. I wouldn't worry about it.

**Comment by nnethercote** [comment]
I'm pretty sure those perf results are just noise, based on (a) how they differ to the pre-merge CI perf run, and (b) similar kinds of results in other PRs lately.

**Pull request merged**